game-app/game-core/src/main/java/games/strategy/triplea/ui/UnitChooser.java

```diff
@@ -648,12 +683,183 @@ public Dimension getMinimumSize() {
 
       @Override
       public Dimension getPreferredSize() {
-        final int width =
-            uiContext.getUnitImageFactory().getUnitImageWidth()
-                * (1 + category.getDependents().size());
-        final int height = uiContext.getUnitImageFactory().getUnitImageHeight();
+        int width = uiContextDecoration.unitImageFactory.getUnitImageWidth()
+                  * (1 + category.getDependents().size());
+
+        if (spaceRequiredForNonWithdrawableIcon) {
+          width += uiContextDecoration.nonWithdrawableImage.getWidth(null) + PADDING;
+        }
+
+        final int height = uiContextDecoration.unitImageFactory.getUnitImageHeight();
         return new Dimension(width, height);
       }
+
+      @VisibleForTesting
+      public ChooserEntry getEmbeddingChooserEntry() {
+        return ChooserEntry.this;
+      }
     }
   }
+
+  /**
+   * once <code>UnitChooser</code> is divided up in a model and a presentation
+   *
+   * (With <code>ChooserEntry</code> being a candidate to go into the model
+   * and <code>UnitChooserEntryIcon</code> being a candidate to go into the presentation,
+   * the method <code>getImage</code> should probably move to <code>UnitChooserEntryIcon</code>.
+   *
+   * It can't be there for the moment, because <code>UnitChooserEntryIcon</code> is an
+   * inner class, and static methods are not supported in inner classes at language level 11.)
+   *
+   * @param category category of the image
+   * @param forceDamaged true iff the image should show a damaged unit
+   * @param uiContext context for loading the image resource
+   * @return the unit image, possibly decorated by the non-withdrawable-icon.
+   */
+
+  @VisibleForTesting
+  public static Image getImage(final UnitCategory category,
+                               final boolean forceDamaged,
+                               final UiContext uiContext) {
+    final var uiContextDecoration = UiContextDecoration.get(uiContext);
+    final Image image = uiContextDecoration.unitImageFactory
```

Comment:
The `unitImageFactory` should be the thing returning the correct image.

`UnitChooser` should be mostly a UI class, it's the image factory's responsibility to find the right images. 

It looks like `uiContextDecoration.getUnitImageWithNonWithdrawableImage(image);` should actually be folded into the unit image factory.

Reply:
I agree. 

>I'd expect basically therefore two factories and some code near where we render the final image that checks if a 'non-withdrawable' should be rendered and if so it would super-impose that image.

I'd like to avoid adding an additional factory and rather move `getUnitImageWithNonWithdrawableImage` into `UnitImageFactory`. Ok?

Reply:
ok

Reply:
I just commited the best solution I could find - expecting a bit of a discussion ;)